lib/Controller/RecordingController.php

```diff
@@ -26,32 +26,96 @@
 
 use InvalidArgumentException;
+use GuzzleHttp\Exception\ConnectException;
 use OCA\Talk\Config;
-use OCA\Talk\Exceptions\UnauthorizedException;
 use OCA\Talk\Service\RecordingService;
-use OCA\Talk\Service\SIPBridgeService;
 use OCP\AppFramework\Http;
 use OCP\AppFramework\Http\DataResponse;
+use OCP\Http\Client\IClientService;
 use OCP\IRequest;
+use Psr\Log\LoggerInterface;
 
 class RecordingController extends AEnvironmentAwareController {
 	public function __construct(
 		string $appName,
 		IRequest $request,
+		private ?string $userId,
 		private Config $talkConfig,
-		private SIPBridgeService $SIPBridgeService,
-		private RecordingService $recordingService
+		private IClientService $clientService,
+		private RecordingService $recordingService,
+		private LoggerInterface $logger
 	) {
 		parent::__construct($appName, $request);
 	}
 
+	public function getWelcomeMessage(int $serverId): DataResponse {
+		$recordingServers = $this->talkConfig->getRecordingServers();
+		if (empty($recordingServers) || !isset($recordingServers[$serverId])) {
+			return new DataResponse([], Http::STATUS_NOT_FOUND);
+		}
+
+		$url = rtrim($recordingServers[$serverId]['server'], '/');
+
+		$client = $this->clientService->newClient();
+		try {
+			$response = $client->get($url . '/api/v1/welcome', [
+				'verify' => (bool) $recordingServers[$serverId]['verify'],
+				'nextcloud' => [
+					'allow_local_address' => true,
+				],
+			]);
+
+			$body = $response->getBody();
+			$data = json_decode($body, true);
+
+			if (!is_array($data)) {
+				return new DataResponse([
+					'error' => 'JSON_INVALID',
+				], Http::STATUS_INTERNAL_SERVER_ERROR);
+			}
+
+			return new DataResponse($data);
+		} catch (ConnectException $e) {
+			return new DataResponse(['error' => 'CAN_NOT_CONNECT'], Http::STATUS_INTERNAL_SERVER_ERROR);
+		} catch (\Exception $e) {
+			return new DataResponse(['error' => $e->getCode()], Http::STATUS_INTERNAL_SERVER_ERROR);
+		}
+	}
+
+	/**
+	 * Check if the current request is coming from an allowed backend.
+	 *
+	 * The backends are sending the custom header "Talk-Recording-Random"
+	 * containing at least 32 bytes random data, and the header
+	 * "Talk-Recording-Checksum", which is the SHA256-HMAC of the random data
+	 * and the body of the request, calculated with the shared secret from the
+	 * configuration.
+	 *
+	 * @param string $data
+	 * @return bool
+	 */
+	private function validateBackendRequest(string $data): bool {
+		$random = $this->request->getHeader('Talk-Recording-Random');
+		if (empty($random) || strlen($random) < 32) {
+			$this->logger->debug("Missing random");
+			return false;
+		}
+		$checksum = $this->request->getHeader('Talk-Recording-Checksum');
+		if (empty($checksum)) {
+			$this->logger->debug("Missing checksum");
+			return false;
+		}
+		$hash = hash_hmac('sha256', $random . $data, $this->talkConfig->getRecordingSecret());
+		return hash_equals($hash, strtolower($checksum));
+	}
+
 	/**
 	 * @PublicPage
```

Comment:
When it's public `userId` can be null.

Also I think recordings require a user at the moment, so I would be fine with limiting it to `@NoAdminRequired`

Reply:
Right, although the recording itself does not require a user uploading the resulting file does, so it should be prevented for guest moderators. I have also adjusted the condition to show the menu item and the documentation.